File: src/server_auto/job_runner.py
"""Simple in-memory job runner."""

import uuid
import time
import threading
from typing import Dict, Any, Optional


# --------------------------------------------------------------
class Job:
    """Represents a background job."""

    def __init__(self, job_type, params):
        """Create a new job."""
        self.job_id = str(uuid.uuid4())[:8]
        self.job_type = job_type
        self.params = params
        self.status = "pending"
        self.result = None
        self.error = None
        self.created_at = time.time()
        self.completed_at = None

    def to_dict(self):
        """Convert job to dictionary."""
        return {
            "job_id": self.job_id,
            "type": self.job_type,
            "status": self.status,
            "result": self.result,
            "error": self.error,
            "created_at": self.created_at,
            "completed_at": self.completed_at,
        }


# --------------------------------------------------------------
class JobRunner:
    """Manages background jobs."""

    def __init__(self):
        """Initialize job runner."""
        self.jobs: Dict[str, Job] = {}

    def start_job(
        self, job_type: str, params: dict
    ) -> Job:
        """Start a new background job."""
        job = Job(job_type, params)
        self.jobs[job.job_id] = job
        job.status = "running"
        thread = threading.Thread(
            target=self._run_job,
            args=(job,),
            daemon=True,
        )
        thread.start()
        return job

    def _run_job(self, job: Job):
        """Execute a job in background."""
        try:
            if job.job_type == "test":
                time.sleep(1)
                job.result = "test completed"
            else:
                job.result = (
                    f"unknown type: "
                    f"{job.job_type}"
                )
            job.status = "completed"
        except Exception as e:
            job.status = "failed"
            job.error = str(e)
        finally:
            job.completed_at = time.time()

    def get_job(self, job_id: str):
        """Get a job by ID."""
        return self.jobs.get(job_id)

    def list_jobs(self):
        """List all jobs."""
        return [
            j.to_dict()
            for j in self.jobs.values()
        ]


runner = JobRunner()

File: server_auto/src/job_runner.rs
/// In-memory background job runner using tokio tasks.
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize)]
pub struct Job {
    pub job_id: String,
    pub job_type: String,
    pub status: String,
    pub result: Option<String>,
    pub error: Option<String>,
    pub created_at: f64,
    pub completed_at: Option<f64>,
}

impl Job {
    fn new(job_type: &str) -> Self {
        let id = &Uuid::new_v4().to_string()[..8];
        Self {
            job_id: id.to_string(),
            job_type: job_type.to_string(),
            status: "pending".to_string(),
            result: None,
            error: None,
            created_at: unix_now(),
            completed_at: None,
        }
    }
}

fn unix_now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

#[derive(Clone)]
pub struct JobState {
    pub jobs: Arc<Mutex<HashMap<String, Job>>>,
}

impl JobState {
    pub fn new() -> Self {
        Self { jobs: Arc::new(Mutex::new(HashMap::new())) }
    }

    pub fn start_job(&self, job_type: &str, _params: Value) -> Job {
        let mut job = Job::new(job_type);
        job.status = "running".to_string();
        let job_id = job.job_id.clone();
        let jtype = job.job_type.clone();
        self.jobs.lock().unwrap().insert(job_id.clone(), job.clone());

        let state = self.clone();
        tokio::spawn(async move {
            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
            let mut jobs = state.jobs.lock().unwrap();
            if let Some(j) = jobs.get_mut(&job_id) {
                if jtype == "test" {
                    j.result = Some("test completed".to_string());
                } else {
                    j.result = Some(format!("unknown type: {jtype}"));
                }
                j.status = "completed".to_string();
                j.completed_at = Some(unix_now());
            }
        });
        job
    }

    pub fn get_job(&self, job_id: &str) -> Option<Job> {
        self.jobs.lock().unwrap().get(job_id).cloned()
    }

    pub fn list_jobs(&self) -> Vec<Job> {
        self.jobs.lock().unwrap().values().cloned().collect()
    }
}
